src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
     }
     Ok(out_dict.into())
 }
+
+struct MisraGriesAlgorithm<'a> {
+    graph: &'a StablePyGraph<Undirected>,
+    colors: DictMap<EdgeIndex, usize>,
+}
+
+impl<'a> MisraGriesAlgorithm<'a> {
+    pub fn new(graph: &'a StablePyGraph<Undirected>) -> Self {
+        let colors: DictMap<EdgeIndex, usize> = DictMap::new();
+        MisraGriesAlgorithm { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: NodeIndex) -> Vec<usize> {
+        let mut used_colors: Vec<usize> = Vec::new();
+        for edge in self.graph.edges(u) {
+            if let Some(color) = self.colors.get(&edge.id()) {
+                used_colors.push(*color);
+            }
+        }
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: NodeIndex) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let mut c: usize = 0;
+        while used_colors.contains(&c) {
+            c += 1;
+        }
+        c
```

Comment:
Just for fun you could do this as an iterator too, but not sure if is any faster. Something like:

```suggestion
        let used_colors = self.get_used_colors(u);
        (0..)
            .position(|color| used_colors.contains(&color))
            .unwrap()
```